### PR TITLE
R&D rewrite.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
- Adds research checking to fabricators.
- Converts all science fabs into actual fabs.
- Adds local network handling for fabricators.
- Converts all R&D designs into fabricator designs that inherit origin_tech from their product.
- Converts destructive analyzer into a standalone machine.
- Adds a design database console that interacts with a network of fabricators to provide tech levels, and the destructive analyzer to obtain tech levels.

### CODE REVIEW

File: code/modules/fabrication/fabricator_industrial.dm [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		MAT_PHORON =    SHEET_MATERIAL_AMOUNT * 100,
+		MAT_URANIUM =   SHEET_MATERIAL_AMOUNT * 100,
+		MAT_DIAMOND =   SHEET_MATERIAL_AMOUNT * 100
+	)
+
+/obj/machinery/fabricator/industrial/do_build(var/datum/fabricator_recipe/recipe, var/amount)
+	. = ..()
+	if(istype(., /atom/movable))
+		var/atom/movable/product = .
+		step(product, EAST)
```
Comment (comma) [inline code/modules/fabrication/fabricator_industrial.dm:27]: maybe check if that turf is accessible so you don't have fabricators shitting into the walls

File: code/modules/fabrication/designs/protolathe/designs_stock.dm [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+/datum/fabricator_recipe/protolathe/stock/basic_sensor
+	path = /obj/item/stock_parts/scanning_module
+
+/datum/fabricator_recipe/protolathe/stock/adv_sensor
+	path = /obj/item/stock_parts/scanning_module/adv
+
+/datum/fabricator_recipe/protolathe/stock/phasic_sensor
+	path = /obj/item/stock_parts/scanning_module/phasic
+
+/datum/fabricator_recipe/protolathe/stock/rped
```
Comment (comma) [inline code/modules/fabrication/designs/protolathe/designs_stock.dm]: this prolly should go into tools with its bluespace bro

File: code/modules/fabrication/fabricator_industrial.dm (hunk 1 above)
Comment (MistakeNot4892) [reply code/modules/fabrication/fabricator_industrial.dm:27]: I am pretty sure `step()` respects turf density/windows/etc.
EDIT: checked, it does.

File: code/modules/fabrication/designs/protolathe/designs_stock.dm (hunk 2 above)
Comment (MistakeNot4892) [reply code/modules/fabrication/designs/protolathe/designs_stock.dm]: Done.

File: code/game/objects/item.dm [hunk 3]
```diff
@@ -204,7 +204,7 @@
 			desc_comp += "<span class='notice'>Testing potentials:</span><BR>"
 			var/list/techlvls = cached_json_decode(origin_tech)
 			for(var/T in techlvls)
-				desc_comp += "Tech: Level [techlvls[T]] [CallTechName(T)] <BR>"
+				desc_comp += "Tech: Level [techlvls[T]] [T] <BR>"
```
Comment (MistakeNot4892) [inline code/game/objects/item.dm]: Need to remember to go fix this.

**Comment by MistakeNot4892** [comment]
This is all in place other than finishing off the design console, probably safe to review.

File: mods/corporate/datum/robolimbs.dm [hunk 4]
```diff
@@ -78,9 +53,54 @@
...
+	company = "Xion"
+	desc = "This limb has a minimalist black and red casing."
+	icon = 'icons/mob/human_races/cyberlimbs/xion/xion_main.dmi'
+
+/datum/robolimb/nanotrasen
+	company = "NanoTrasen"
+	desc = "This limb is made from a cheap polymer."
+	icon = 'icons/mob/human_races/cyberlimbs/nanotrasen/nanotrasen_main.dmi'
+
+#define DEFINE_ROBOLIMB_DESIGNS(MODEL_PATH, MODEL_ID, MODEL_NAME)         \
```
Comment (comma) [inline mods/corporate/datum/robolimbs.dm]: I think this macro should go into the main robolimb file, not mod's one. Not really corp specific is it.

File: mods/corporate/datum/robolimbs.dm [hunk 5]
```diff
@@ -78,9 +53,54 @@
...
+/datum/fabricator_recipe/robotics/prosthetic/model_##MODEL_ID {           \
+	category = "MODEL_NAME Prosthetics";                                  \
```
Comment (comma) [inline mods/corporate/datum/robolimbs.dm]: also probably should be `category = MODEL_NAME + "Prosthetics"`

File: code/modules/fabrication/designs/protolathe/designs_misc.dm [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+/datum/fabricator_recipe/protolathe/misc/bag_holding
+	path = /obj/item/storage/backpack/holding
+
+/datum/fabricator_recipe/protolathe/misc/dufflebag_holding
+	path = /obj/item/storage/backpack/holding/duffle
+
+/datum/fabricator_recipe/protolathe/misc/beaker
+	path = /obj/item/chems/glass/beaker/noreact
+
+/datum/fabricator_recipe/protolathe/beaker/get_product_name()
```
Comment (comma) [inline code/modules/fabrication/designs/protolathe/designs_misc.dm]: Should be `/datum/fabricator_recipe/protolathe/misc/beaker`? That's where both bluespace and noreact live

**Review by comma — DISMISSED** [review]

**Comment by MistakeNot4892** [comment]
Ready for review.

**Review by comma — DISMISSED** [review]

File: code/modules/research/design_database_analyzer.dm [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+		visible_message("\The [user] unloads \the [src]'s material hopper.")
+		for(var/mat in dump_matter)
+			var/material/M = SSmaterials.get_material_datum(mat)
+			M.place_sheet(loc, dump_matter[mat])
+			cached_materials[mat] -= dump_matter[mat] * SHEET_MATERIAL_AMOUNT
+			if(cached_materials[mat] <= 0)
+				cached_materials -= mat
+		return TRUE
+
+	. = ..()
```
Comment (afterthought2) [inline code/modules/research/design_database_analyzer.dm:74]: Use `physical_attack_hand` and `interface_interact` instead for machines.

File: code/modules/research/design_database_analyzer.dm [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	anchored = TRUE
+
+	var/local_network_tag
+	var/busy = FALSE
+	var/obj/item/loaded_item = null
+	var/material_return_modifier = 0
+	var/list/cached_materials = list()
+
+/obj/machinery/destructive_analyzer/examine(mob/user)
+	. = ..()
```
Comment (afterthought2) [inline code/modules/research/design_database_analyzer.dm]: Pointless override?

File: code/modules/research/design_database.dm [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+		TECH_ESOTERIC =    0
+	)
+
+	var/need_disk_operation = FALSE
+	var/obj/item/disk/tech_disk/disk
+	var/sync_policy = SYNC_PULL_NETWORK|SYNC_PULL_DISK
+
+/obj/machinery/design_database/CanUseTopic(var/user)
+	if(istype(user, /mob/living/silicon) || (Adjacent(user)))
+		return STATUS_INTERACTIVE
```
Comment (afterthought2) [inline code/modules/research/design_database.dm]: This is not a desirable override. If there's something in particular you're trying to accomplish with it, ask me.

File: code/modules/research/design_database_analyzer.dm (hunk 7 above)
Comment (MistakeNot4892) [reply code/modules/research/design_database_analyzer.dm:74]: Swapped in `physical_attack_hand`, not sure where you want `interface_interact` used as I used it for the other machines.

**Review by afterthought2 — DISMISSED** [review]

**Review by afterthought2 — APPROVED** [review]

**Pull request merged**